### PR TITLE
MODINV-748 - Link update: Update instance according to srs.bib event

### PR DESCRIPTION
**Pull request by sherzodn**
## Purpose
Link update: Update instance according to srs.bib event

## Approach

1. Introduce a new verticle that will init consumer for `srs.marc-bib` topic
2. Register verticle in `Launcher`
3. Introduce a handler for a consumer. It should:

- Retrieve mapping rules and mapping parameters for marc-bib (GET /mapping-metadata/type/marc-bib)
- Use local cache to store mapping metadata (MappingMetadataCache), use jobId from message
- Use InstanceUpdateDelegate to update instance record in mod-inventory-storage

### Changes checklist
- [ ] API paths, methods, request or response bodies changed, added, or removed
- [ ] Database schema changes
- [ ] Interface versions changes
- [ ] Interface dependencies added, or removed
- [ ] Permissions changed, added, or removed

## Learnings
https://issues.folio.org/browse/MODINV-748

### CODE REVIEW

File: src/main/java/org/folio/inventory/MarcBibUpdateConsumerVerticle.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+    return KafkaConsumerWrapper.<String, String>builder()
+      .context(context)
+      .vertx(vertx)
+      .kafkaConfig(kafkaConfig)
+      .loadLimit(loadLimit)
+      .globalLoadSensor(GLOBAL_LOAD_SENSOR)
+      .subscriptionDefinition(subscriptionDefinition)
+      .build();
+  }
```
Comment (psmagin) [inline src/main/java/org/folio/inventory/MarcBibUpdateConsumerVerticle.java:64]: The full topic name will have the format `<env>.<tenant>.srs.marc-bib`.
But in this method, you defined it as `<env>.<namespace>.<tenant>.srs.marc-bib`

File: src/main/java/org/folio/inventory/dataimport/consumers/MarcBibUpdateKafkaHandler.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+  public MarcBibUpdateKafkaHandler(InstanceUpdateDelegate instanceUpdateDelegate, MappingMetadataCache mappingMetadataCache) {
+    this.instanceUpdateDelegate = instanceUpdateDelegate;
+    this.mappingMetadataCache = mappingMetadataCache;
+  }
+
+  @Override
+  public Future<String> handle(KafkaConsumerRecord<String, String> record) {
+    try {
+      Promise<String> promise = Promise.promise();
+      Event event = OBJECT_MAPPER.readValue(record.value(), Event.class);
```
Comment (psmagin) [inline src/main/java/org/folio/inventory/dataimport/consumers/MarcBibUpdateKafkaHandler.java:54]: Our message will not be wrapped by Event class

File: src/main/java/org/folio/inventory/dataimport/consumers/MarcBibUpdateKafkaHandler.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+      @SuppressWarnings("unchecked")
+      HashMap<String, String> eventPayload = OBJECT_MAPPER.readValue(event.getEventPayload(), HashMap.class);
```
Comment (psmagin) [inline src/main/java/org/folio/inventory/dataimport/consumers/MarcBibUpdateKafkaHandler.java:56]: You should create a class that will describe the format of our event and then deserialize it there.

File: src/main/java/org/folio/inventory/dataimport/consumers/MarcBibUpdateKafkaHandler.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+      Map<String, String> headersMap = KafkaHeaderUtils.kafkaHeadersToMap(record.headers());
+      String recordId = headersMap.get(RECORD_ID_HEADER);
+      String chunkId = headersMap.get(CHUNK_ID_HEADER);
```
Comment (psmagin) [inline src/main/java/org/folio/inventory/dataimport/consumers/MarcBibUpdateKafkaHandler.java:59]: We don't have these fields in message.

File: src/main/java/org/folio/inventory/dataimport/consumers/MarcBibUpdateKafkaHandler.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+      String jobExecutionId = eventPayload.get(JOB_ID);
+      LOGGER.info("Event payload has been received with event type: {}, recordId: {} by jobExecution: {} and chunkId: {}", event.getEventType(), recordId, jobExecutionId, chunkId);
+
+      Context context = EventHandlingUtil.constructContext(headersMap.get(OKAPI_TENANT_HEADER), headersMap.get(OKAPI_TOKEN_HEADER), headersMap.get(OKAPI_URL_HEADER));
```
Comment (psmagin) [inline src/main/java/org/folio/inventory/dataimport/consumers/MarcBibUpdateKafkaHandler.java:63]: Tenant is a part of a message, it's not in a header.

File: src/main/java/org/folio/inventory/dataimport/consumers/MarcBibUpdateKafkaHandler.java [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+      mappingMetadataCache.getByRecordType(jobExecutionId, context, MARC_BIB_RECORD_TYPE)
+        .map(metadataOptional -> metadataOptional.orElseThrow(() ->
+          new EventProcessingException(format(MAPPING_METADATA_NOT_FOUND_MSG, jobExecutionId))))
+        .onSuccess(mappingMetadataDto -> ensureEventPayloadWithMappingMetadata(eventPayload, mappingMetadataDto))
+        .compose(v -> instanceUpdateDelegate.handle(eventPayload, marcBibRecord, context))
+        .onComplete(ar -> {
+          if (ar.succeeded()) {
+            promise.complete(record.key());
+          } else {
+            LOGGER.error("Failed to update marc-bib by jobExecutionId {}:{}", jobExecutionId, ar.cause());
```
Comment (psmagin) [inline src/main/java/org/folio/inventory/dataimport/consumers/MarcBibUpdateKafkaHandler.java:75]:
```suggestion
            LOGGER.error("Failed to update instance by jobId {}:{}", jobExecutionId, ar.cause());
```

File: src/main/java/org/folio/inventory/dataimport/consumers/MarcBibUpdateKafkaHandler.java [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+  public Future<String> handle(KafkaConsumerRecord<String, String> record) {
+    try {
+      Promise<String> promise = Promise.promise();
+      Event event = OBJECT_MAPPER.readValue(record.value(), Event.class);
+      @SuppressWarnings("unchecked")
+      HashMap<String, String> eventPayload = OBJECT_MAPPER.readValue(event.getEventPayload(), HashMap.class);
+      Map<String, String> headersMap = KafkaHeaderUtils.kafkaHeadersToMap(record.headers());
+      String recordId = headersMap.get(RECORD_ID_HEADER);
+      String chunkId = headersMap.get(CHUNK_ID_HEADER);
+      String jobExecutionId = eventPayload.get(JOB_ID);
```
Comment (psmagin) [inline src/main/java/org/folio/inventory/dataimport/consumers/MarcBibUpdateKafkaHandler.java:60]:
```suggestion
      String jobId = eventPayload.get(JOB_ID);
```

File: src/main/java/org/folio/inventory/Launcher.java [hunk 8]
```diff
@@ -87,20 +89,25 @@ private static void startConsumerVerticles(Map<String, Object> consumerVerticles
     int dataImportConsumerVerticleNumber = Integer.parseInt(System.getenv().getOrDefault(DATA_IMPORT_CONSUMER_VERTICLE_INSTANCES_NUMBER_CONFIG, "3"));
     int instanceHridSetConsumerVerticleNumber = Integer.parseInt(System.getenv().getOrDefault(MARC_BIB_INSTANCE_HRID_SET_CONSUMER_VERTICLE_INSTANCES_NUMBER_CONFIG, "3"));
     int quickMarcConsumerVerticleNumber = Integer.parseInt(System.getenv().getOrDefault(QUICK_MARC_CONSUMER_VERTICLE_INSTANCES_NUMBER_CONFIG, "1"));
+    int marcBibUpdateConsumerVerticleNumber = Integer.parseInt(System.getenv().getOrDefault(MARC_BIB_UPDATE_CONSUMER_VERTICLE_INSTANCES_NUMBER_CONFIG, "1"));
```
Comment (psmagin) [inline src/main/java/org/folio/inventory/Launcher.java:92]: Let's make it 3

**Review by psmagin — CHANGES_REQUESTED** [review]

File: src/main/java/org/folio/inventory/MarcBibUpdateConsumerVerticle.java (hunk 1 above)
Comment (sherzodn) [reply src/main/java/org/folio/inventory/MarcBibUpdateConsumerVerticle.java]: done

**Review by Shans-Kaluhin — APPROVED** [review]
Pavlo's comments were complied

**Review by psmagin — APPROVED** [review]

**Pull request merged**